src/asm/charmap.cpp

```diff
@@ -14,178 +17,127 @@
 #include "asm/output.hpp"
 #include "asm/warning.hpp"
 
-#include "hashmap.hpp"
 #include "util.hpp"
 
 // Charmaps are stored using a structure known as "trie".
 // Essentially a tree, where each nodes stores a single character's worth of info:
 // whether there exists a mapping that ends at the current character,
-struct Charnode {
+struct CharmapNode {
 	bool isTerminal; // Whether there exists a mapping that ends here
 	uint8_t value; // If the above is true, its corresponding value
 	// This MUST be indexes and not pointers, because pointers get invalidated by `realloc`!
 	size_t next[255]; // Indexes of where to go next, 0 = nowhere
 };
 
 struct Charmap {
-	char *name;
-	std::vector<struct Charnode> *nodes; // first node is reserved for the root node
+	std::string name;
+	std::vector<struct CharmapNode> nodes; // first node is reserved for the root node
 };
 
-static HashMap charmaps;
+static std::map<std::string, struct Charmap> charmaps;
```

Comment:
I don't like that charmaps are not tries anymore, because then applying the charmap is a lot more expensive (how do you do leftmost-**longest** without rescanning logic? Tries can backtrack much more easily).

Reply:
Aren't they still tries though? The object you're commenting on is used for looking up charmaps by name, not for applying the charmap.

Reply:
Yes, charmaps are still a 255-ary trie. 

```c++
// Charmaps are stored using a structure known as "trie".
// Essentially a tree, where each nodes stores a single character's worth of info:
// whether there exists a mapping that ends at the current character,
struct CharmapNode {
	bool isTerminal; // Whether there exists a mapping that ends here
	uint8_t value; // If the above is true, its corresponding value
	// This MUST be indexes and not pointers, because pointers get invalidated by `realloc`!
	size_t next[255]; // Indexes of where to go next, 0 = nowhere
};

struct Charmap {
	std::string name;
	std::vector<struct CharmapNode> nodes; // first node is reserved for the root node
};
```

Each charmap owns its `nodes`, and each node has an array of 255 indexes into `nodes` by which to descend the trie.